libs/imageio/src/BlockCompression.cpp

```diff
@@ -0,0 +1,249 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <imageio/BlockCompression.h>
+
+#include <image/ImageOps.h>
+
+#include <cmath>
+
+#include <astcenc.h>
+
+using namespace image;
+
+using std::string;
+
+namespace image {
+
+AstcTexture astcCompress(const LinearImage& original, AstcConfig config) {
+
+    // If this is the first time, initialize the ARM encoder tables.
+
+    static bool first = true;
+    if (first) {
+        test_inappropriate_extended_precision();
+        prepare_angular_tables();
+        build_quantization_mode_table();
+        first = false;
+    }
+
+    // Create an input image for the ARM encoder in a format that it can consume.
+    // It expects four-channel data, so we extend or curtail the channel count in a reasonable way.
+    // The encoder can take half-floats or bytes, but we always give it half-floats.
+
+    LinearImage source = original;
+    const uint32_t width = source.getWidth();
+    const uint32_t height = source.getHeight();
+    auto createEmptyImage = [width, height](float value) {
+        auto result = LinearImage(width, height, 1);
+        float* pixels = result.getPixelRef(0, 0);
+        std::fill(pixels, pixels + width * height, value);
+        return result;
+    };
+    switch (source.getChannels()) {
+        case 4: break;
+        case 1: {
+            auto l = image::extractChannel(source, 0);
+            auto a = createEmptyImage(1.0f);
+            source = image::combineChannels({l, l, l, a});
+        }
+        case 2: {
+            auto l = image::extractChannel(source, 0);
+            auto a = image::extractChannel(source, 1);
+            source = image::combineChannels({l, l, l, a});
+        }
+        case 3: {
+            auto r = image::extractChannel(source, 0);
+            auto g = image::extractChannel(source, 1);
+            auto b = image::extractChannel(source, 2);
+            auto a = createEmptyImage(1.0f);
+            source = image::combineChannels({r, g, b, a});
+        }
+        default: {
+            auto r = image::extractChannel(source, 0);
+            auto g = image::extractChannel(source, 1);
+            auto b = image::extractChannel(source, 2);
+            auto a = image::extractChannel(source, 3);
+            source = image::combineChannels({r, g, b, a});
+        }
+    }
+
+    astc_codec_image* input_image = allocate_image(16, width, height, 1, 0);
+    constexpr bool flip = false;
+    auto imagedata16 = input_image->imagedata16[0];
+    for (int y = 0; y < height; y++) {
+        int y_dst = y;
+        int y_src = flip ? (height - y - 1) : y;
```

Comment:
`flip` is always `false`

Reply:
Removed.  I thought maybe we'd need it in the future but I agree this was a bit sloppy.